mempool/src/pool.rs

```diff
@@ -0,0 +1,1007 @@
+
+const MEMPOOL_MAX_TXS: usize = 1_000_000;
+
+pub trait Mempool<C> {
+    fn create(chain_state: C) -> Self;
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError>;
+    fn get_all(&self) -> Vec<&Transaction>;
+    fn contains_transaction(&self, tx: &Id<Transaction>) -> bool;
+    fn drop_transaction(&mut self, tx: &Id<Transaction>);
+    fn new_tip_set(&mut self) -> Result<(), MempoolError>;
+}
+
+pub trait ChainState {
+    fn contains_outpoint(&self, outpoint: &OutPoint) -> bool;
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct TxMempoolEntry {
+    tx: Transaction,
+    parents: BTreeSet<Rc<TxMempoolEntry>>,
+}
+
+impl TxMempoolEntry {
+    fn new(tx: Transaction, pool: &MempoolStore) -> Self {
+        let parents = tx
+            .get_inputs()
+            .iter()
+            .filter_map(|input| pool.txs_by_id.get(&input.get_outpoint().get_tx_id().get()))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+
+        Self { tx, parents }
+    }
+    //TODO this should really be sum of inputs minus sum of outputs
+    //But we don't yet have a way of summing of the inputs
+    //TODO handle the error bettter
+    fn get_fee(&self) -> Amount {
+        self.tx
+            .get_outputs()
+            .iter()
+            .map(|output| output.get_value())
+            .sum::<Option<_>>()
+            .expect("fee overflow")
+    }
+
+    fn is_replaceable(&self) -> bool {
+        self.tx.is_replaceable()
+            || self.unconfirmed_ancestors().iter().any(|ancestor| ancestor.tx.is_replaceable())
+    }
+
+    fn unconfirmed_ancestors(&self) -> BTreeSet<Rc<TxMempoolEntry>> {
+        let mut visited = BTreeSet::new();
+        self.unconfirmed_ancestors_inner(&mut visited);
+        visited
+    }
+
+    fn unconfirmed_ancestors_inner(&self, visited: &mut BTreeSet<Rc<TxMempoolEntry>>) {
+        for parent in self.parents.iter() {
+            if visited.contains(parent) {
+                break;
+            } else {
+                visited.insert(Rc::clone(parent));
+                parent.unconfirmed_ancestors_inner(visited);
+            }
+        }
+    }
+
+    fn compare_fees(&self, other: &TxMempoolEntry) -> Ordering {
+        // We want the comparison to be such that a collection is sorted if its elements are ordered in
+        // descending order
+        other.get_fee().cmp(&self.get_fee())
+    }
+}
+
+impl PartialOrd for TxMempoolEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.compare_fees(other))
+    }
+}
+
+impl Ord for TxMempoolEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.compare_fees(other)
+    }
+}
+
+#[derive(Debug)]
+pub struct MempoolImpl<C: ChainState> {
+    store: MempoolStore,
+    chain_state: C,
+}
+
+#[derive(Debug)]
+struct MempoolStore {
+    txs_by_id: HashMap<H256, Rc<TxMempoolEntry>>,
+    txs_by_fee: BTreeSet<Rc<TxMempoolEntry>>,
+    spender_txs: HashMap<Id<OutPoint>, Rc<TxMempoolEntry>>,
```

Comment:
How about just using a BTreeMap? Isn't that better than attempting to hash OutPoint?

Reply:
@TheQuantumPhysicist I thought about it, but not sure what the sensible way  would be to implement `Ord` for `OutPoint` (required for BTreeMap).

Reply:
Isn't an OutPoint just a hash and an index? Why not just Ord for hash then index?

Reply:
You're right. The `OutPoint` *is* a kind of Id.
I've gotten rid of the `Idable for Outpoint`  and changed the code to use BTreeSets instead.